controllers/result.go

```diff
@@ -133,64 +134,57 @@ func GetResult(ctx context.Context, w http.ResponseWriter, r *http.Request) (int
 		return http.StatusUnauthorized, nil
 	}
 
-	if p.UserKey.Parent() != nil {
-		json.NewEncoder(w).Encode(result.Key(resultKey))
-		return http.StatusOK, nil
+	if p.UserKey.Parent() != nil && !util.HasParent(resultKey, p.UserKey) {
+		return http.StatusUnauthorized, nil
 	}
 
-	if !util.HasParent(resultKey, p.UserKey) {
-		return http.StatusUnauthorized, nil
+	if result.Finished == (time.Time{}) {
+		return createFinalResult(ctx, w, resultKey, result)
 	}
-	return createFinalResult(ctx, w, resultKey, result)
+
+	json.NewEncoder(w).Encode(result.Key(resultKey))
+	return http.StatusOK, nil
 }
 
-func createFinalResult(ctx context.Context, w http.ResponseWriter, resultKey *datastore.Key, result model.Result) (int, error) {
+func createFinalResult(ctx context.Context, w http.ResponseWriter, resultKey *datastore.Key, result model.Result) (status int, err error) {
 	go computeFinalScore(ctx, result)
 
 	var challenge model.Challenge
 	if err := datastore.Get(ctx, result.Challenge, &challenge); err != nil {
-		return http.StatusInternalServerError, nil
+		return http.StatusInternalServerError, err
 	}
 
-	var taskKey *datastore.Key
-	for _, taskKey = range challenge.Tasks {
-		switch taskKey.Kind() {
-		case model.CodeTaskKind:
-			var submissions model.CodeSubmissions
-			keys, err := doQuery(ctx, model.NewQueryForCodeSubmission(), resultKey, taskKey, submissions)
-			if err != nil {
-				return http.StatusInternalServerError, nil
-			}
-			if len(keys) == 0 {
-				// Most likely the authenticated user called this endpoint
-				// before finishing the challenge
-				return http.StatusUnauthorized, nil
-			}
-			result.FinalSubmissions = append(result.FinalSubmissions, keys[0])
-		//TODO(pbochis, vbalan, flowlo): Add more cases when more task kinds are added.
-		default:
-			return http.StatusBadRequest, errors.New("Unknown submission kind.")
+	result.Finished = time.Now()
+
+	for i, taskKey := range challenge.Tasks {
+		var key *datastore.Key
+		if key, err = getLastSubmission(ctx, resultKey, taskKey); err != nil {
+			return http.StatusInternalServerError, err
 		}
+		result.FinalSubmissions[i] = key
 	}
-	key, err := result.Save(ctx, resultKey)
-	if err != nil {
+
+	if _, err = result.Save(ctx, resultKey); err != nil {
 		return http.StatusInternalServerError, err
 	}
-	json.NewEncoder(w).Encode(result.Key(key))
-	return http.StatusOK, nil
+	json.NewEncoder(w).Encode(result.Key(resultKey))
+	return
 }
 
-func doQuery(ctx context.Context, query *datastore.Query, resultKey, taskKey *datastore.Key, dst interface{}) ([]*datastore.Key, error) {
-	keys, err := query.
+func getLastSubmission(ctx context.Context, resultKey, taskKey *datastore.Key) (*datastore.Key, error) {
+	keys, err := model.NewQueryForCodeSubmission().
```

Comment:
The previous implementation was agnostic to any kind of submission. Why did you drop that?